### ISSUE TITLE
Option to un-center offset (origin) on collision shape rectangle

### ISSUE THREAD

**Issue by pajcek11** [godotengine/godot-proposals#1170]
**Describe the project you are working on:**
Trying to have a node with a Sprite and Collision box/shape

**Describe the problem or limitation you are having in your project:**
The sprite's "origin" can be set to un-centered (top-left) as is the standard in many programs, but the collision shape allows for only the origin being "centered", messing everything up because now you can't just overlay the sprite and shape one over the other because they have two completely different origin points. Besides, what happens with pixel-perfection if the sprite has a size not divisible by 2? Where does then the center lie? That's why top-left is standard and superior.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
Put a checkbox for "centered" in the collision shape, so people can un-check it just like they can on the sprite

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
Collision shape rectangle should have some kind of centered property already, so just making the checkbox should be easy, if I am not wrong.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
It will most definitely be used often because it is essential and absolutely logical to have

**Is there a reason why this should be core and not an add-on in the asset library?:**
This should be an easy fix, and a logical addition. The current system is flawed for not having this.

**Comment by Calinou** [comment]
Related to #452.

> That's why top-left is standard and superior.

Don't be so sure :slightly_smiling_face: 
Some software has moved from top-left origins to centered origins for various reasons. In many cases, it's more logical to extend from the center than from the top-left corner.

**Comment by pajcek11** [comment]
I really do hope that something is done about this, and that people will support it. 
I wholeheartedly support issue 452.

>Godot allows floating-point coordinates in 2D

I don't allow them though because they don't exist and make no sense.

>It's easy to implement, but it also means having to add and document yet another property (and therefore bloating the Inspector). Everything has a cost, even if it's not obvious at first.

That's my kind of work-ethic ;P We'll never accomplish anything thinking like that.

________________

Anyway, could you suggest some workaround for me in the meantime? It appears that the offset functionality isn't part of Node2D, so those properties don't even exist in the collision shapes? Could we inherit it somehow and then export the centered property? The goal is to be able to position collision shapes via code or editor, without using floating points, but real tangible pixel coordinates.

Tnx!

**Comment by KoBeWi** [comment]
> Anyway, could you suggest some workaround for me in the meantime?

I made a PR for #452 and I actually use it already in my custom Godot build (it's the same as 3.2.2 stable plus some bonus stuff), you could use it in the meantime: https://ufile.io/icpr90x8 (also you can test if #452 solves your problem)

**Comment by Shadowblitz16** [comment]
this needs to be added.
I need a way to align my tiles to a grid and tilemaps don't cut it
tilemaps don't support moving objects and collision logic

**Comment by Georges-SR** [comment]
Any news on this important feature ?

**Comment by Calinou** [comment]
> Any news on this convenient feature ? I wholeheartedly support it, as many others I'm sure.

To my knowledge, nobody is currently working on implementing this. This feature doesn't seem to have a lot of community support, and would come with a fair bit of added complexity on its own. Also, we can have editor helpers to resize shapes from their top-left (like we have in 3D already).

**Comment by simonlow0210** [comment]
I would also suggest to add some sort of support of RectangleShape2D for allowing to choose an anchor point similar to Control Nodes of ColorRect or TextureRect.

For example, option of "grow from bottom / anchor from bottom" or something similar would make it much easier to animate a character with collision shape that changes size every frame. Right now it is a big hassle to resize the collision shape with pure coding and needing to change the position as well.

This is especially true for action or fighting games with sprites of different heights.

**Comment by neomonkey** [comment]
I am voting on this one. I just need basic functionality to find mouse enter and exit. I need to have an area2D and collisionshape2D to get the thing done, but it won't even snap with my sprite unless I do some extra code. Basically it is not a big thing on my end, but I agree with the comment that top left is pretty common to make calculations snap easily and release quite substantial amount of calculation code needed. I think @simonlow0210 comment is really good, because on my end I just need a way to track input in an area, and it is a bit annoying to calculate with the current system.

**Comment by ThePsydevil888** [comment]
Yes please add this. It's not a thing that can easily be worked around, or at least I don't see an easy way to do so. The lack of an offset/pivot functionality on a CollisionShape2D, just like the -

**Comment by yaltaki** [comment]
Bump.

It's really not that hard of a feature to add, and the fact that some object types have it while CollisionShape doesn't isn't being lightweight, it's just being inconsistent. And it's such a core feature you'd expect a 2D engine to have that I can't imagine anyone calling something like this bloat or unnecessary.

This issue has been open for 5 years!!! Surely it doesn't need so much deliberation!?

**Comment by Calinou** [comment]
Now that https://github.com/godotengine/godot-proposals/issues/452 has been implemented in 3D, I think the way to go is to implement a similar mechanism in 2D. This way, only the editor side has to be modified, which is simpler from a physics engine perspective.

**Comment by KoBeWi** [comment]
It already exists in 2D.

**Comment by Lamby777** [comment]
> Godot allows floating-point coordinates in 2D (even if pixel snap is enabled), so I don't think this is much of an issue.

This just isn't true...? Unless I'm doing something wrong, the blue box doesn't cooperate when `snap_2d_transforms_to_pixel` is turned on and the box position is 0.5 (attached pic). As soon as snap is turned off, the blue box moves to the expected position between the orange dots.

<img width="219" height="132" alt="Image" src="https://github.com/user-attachments/assets/fdde4809-450a-4fd2-b488-69cefb86da06" />

In this case, even just being able to offset its origin by 0.5 in the collision shape settings would solve everything. Instead, I'm forced to use manually specified 4-point collision polygons... (not to mention capsule, circle, etc aren't even possible now)

<img width="306" height="164" alt="Image" src="https://github.com/user-attachments/assets/30e7c7fc-08eb-489e-86cd-0f7a73c1c894" />

**Comment by Calinou** [comment]
> This just isn't true...? Unless I'm doing something wrong, the blue box doesn't cooperate when `snap_2d_transforms_to_pixel` is turned on and the box position is 0.5 (attached pic). As soon as snap is turned off, the blue box moves to the expected position between the orange dots.

What I was saying is that the floating-point coordinates remain floating-points even if pixel snap is enabled. Pixel snap is purely visual; it will never affect the outcome of physics simulations or game logic.

**Comment by BaconEggsRL** [comment]
> Right now it is a big hassle to resize the collision shape with pure coding and needing to change the position as well.

Could someone share how you would do this in code? If there's a workaround I'd like to try it.

I agree this feature should be added. It's possible to tween/animate the CollisionShape2D using the size x & y properties.
However the origin being stuck at the center is somewhat limiting.
Also tweening scale does not seems to work like it does on other nodes.
